data-node/src/main/java/org/graylog/datanode/docs/ConfigurationDocsGenerator.java

```diff
@@ -0,0 +1,138 @@
+import java.util.Locale;
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+public class ConfigurationDocsGenerator {
+
+    /**
+     * This class is linked from the datanode pom.xml and generates conf.example and csv documentation.
+     */
+    public static void main(String[] args) throws IOException {
+        final ConfigurationDocsGenerator generator = new ConfigurationDocsGenerator();
+        generator.generateDocumentation(parseDocumentationFormat(args), ConfigurationDocsGenerator::getDatanodeConfigurationBeans);
+    }
+
+    @Nonnull
+    private static DocumentationFormat parseDocumentationFormat(String[] args) {
+        if (args.length != 2) {
+            throw new IllegalArgumentException("This command needs two arguments - a format and file path. For example" + "csv ${project.build.directory}/configuration-documentation.csv");
+        }
+        final String format = args[0].toLowerCase(Locale.ROOT).trim();
+        final String file = args[1];
+        return new DocumentationFormat(format, file);
+    }
+
+    protected void generateDocumentation(DocumentationFormat format, Supplier<List<Object>> configurationBeans) throws IOException {
+        final List<ConfigurationEntry> configuration = detectConfigurationFields(configurationBeans);
+        try (final DocsPrinter writer = createWriter(format)) {
+            writer.writeHeader();
+
+            for (ConfigurationEntry f : configuration) {
+                writer.writeField(f);
+            }
+        }
+    }
+
+    private DocsPrinter createWriter(DocumentationFormat format) throws IOException {
+        final FileWriter fileWriter = new FileWriter(format.outputFile(), StandardCharsets.UTF_8);
+        return switch (format.format()) {
+            case "csv" -> new CsvDocsPrinter(fileWriter);
+            case "conf" -> new ConfigFileDocsPrinter(fileWriter);
+            default -> throw new IllegalArgumentException("Unsupported format " + format.format());
+        };
+    }
+
+    /**
+     * Collects all configuration options from all available configuration beans.
+     */
+    private List<ConfigurationEntry> detectConfigurationFields(Supplier<List<Object>> configurationBeans) {
+        return configurationBeans.get()
+                .stream()
+                .flatMap(ConfigurationDocsGenerator::beanToConfigEntries)
+                .sorted(Comparator.comparing(ConfigurationEntry::isPriority).reversed())
+                .toList();
+    }
+
+    @Nonnull
+    private static Stream<ConfigurationEntry> beanToConfigEntries(Object configurationBean) {
+        return Arrays.stream(configurationBean.getClass().getDeclaredFields())
+                .filter(f -> f.isAnnotationPresent(Parameter.class))
+                .filter(ConfigurationDocsGenerator::isPublicFacing)
+                .map(f -> toConfigurationEntry(f, configurationBean));
+    }
+
+    private static List<Object> getDatanodeConfigurationBeans() {
```

Comment:
Maybe we should already move that out of the generator to be able to move the generator to a server package more easily?

Reply:
Do you have any simple idea how/where to get this?